File: Test/javascript/single_agent.js
/**
 * single_agent.js  –  Use Case 1
 * ================================
 * One script controls one character in the game.
 * The Unity display follows this agent's point-of-view.
 *
 * Usage:
 *   node single_agent.js
 *
 * Requirements:
 *   npm install ws        (already in package.json)
 */

'use strict';

const WebSocket = require('ws');

const UNITY_WS_URL = 'ws://localhost:8080/agent';
const AGENT_ID     = 'agent_01';

// ---------------------------------------------------------------------------
// Low-level helpers
// ---------------------------------------------------------------------------

function makeCommand(commandType, data = {}, agentId = AGENT_ID) {
  return JSON.stringify({
    commandType,
    data,
    agentId,
    timestamp: Date.now() / 1000,
  });
}

function send(ws, commandType, data = {}, agentId = AGENT_ID) {
  ws.send(makeCommand(commandType, data, agentId));
  console.log(`[${agentId}] → ${commandType}`, data);
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// ---------------------------------------------------------------------------
// Wall-collision / stuck detector
// ---------------------------------------------------------------------------

const STUCK_FRAMES = 3;
const POS_EPSILON  = 0.02;  // metres

const stuck = {
  isMoving:   false,
  lastPos:    null,
  stuckCount: 0,

  /** Call with msg.player.position each time a broadcast arrives. */
  onPosition(pos) {
    if (!this.isMoving) {
      this.stuckCount = 0;
      this.lastPos    = null;
      return false;
    }
    if (this.lastPos) {
      const same =
        Math.abs(pos.x - this.lastPos.x) < POS_EPSILON &&
        Math.abs(pos.y - this.lastPos.y) < POS_EPSILON &&
        Math.abs(pos.z - this.lastPos.z) < POS_EPSILON;
      this.stuckCount = same ? this.stuckCount + 1 : 0;
    }
    this.lastPos = { ...pos };
    return this.stuckCount >= STUCK_FRAMES;
  },
};

// ---------------------------------------------------------------------------
// High-level actions
// ---------------------------------------------------------------------------

const move = (ws, x, z) => {
  stuck.isMoving = true; stuck.stuckCount = 0;
  send(ws, 'MOVE', { x, z });
};
const stop = (ws) => {
  stuck.isMoving = false;
  send(ws, 'STOP', {});
};
const look         = (ws, pitch, yaw)=> send(ws, 'LOOK',         { pitch, yaw });
const shoot        = (ws, active, duration = 0) => send(ws, 'SHOOT', { active, duration });
const reloadWeapon = (ws)            => send(ws, 'RELOAD',       {});
const switchWeapon = (ws, slot)      => send(ws, 'SWITCH_WEAPON',{ weaponIndex: slot });
const aim          = (ws, active)    => send(ws, 'AIM',          { active });
const setView      = (ws, targetId = AGENT_ID) =>
  send(ws, 'SET_VIEW', { viewTargetAgentId: targetId });

// ---------------------------------------------------------------------------
// Game-state listener
// ---------------------------------------------------------------------------

function onMessage(raw, ws) {
  try {
    const msg = JSON.parse(raw);
    if (msg.type === 'welcome') {
      console.log(`[${AGENT_ID}] Connected – session ${msg.sessionId}`);
      return;
    }
    if (msg.player) {
      const p = msg.player;
      console.log(
        `[${AGENT_ID}] HP:${p.health}/${p.maxHealth} ` +
        `Ammo:${p.currentAmmo}/${p.maxAmmo} ` +
        `State:${p.movementState} ` +
        `Enemies:${(msg.enemies || []).length}`
      );
      if (p.position && stuck.onPosition(p.position)) {
        console.log(`[${AGENT_ID}] Wall detected — auto-stop.`);
        stuck.isMoving = false;
        send(ws, 'STOP', {});
      }
    }
  } catch (e) {
    console.error(`[${AGENT_ID}] Parse error:`, e.message);
  }
}

// ---------------------------------------------------------------------------
// Main scenario
// ---------------------------------------------------------------------------

async function main() {
  console.log(`[${AGENT_ID}] Connecting to ${UNITY_WS_URL} ...`);

  const ws = new WebSocket(UNITY_WS_URL);

  await new Promise((resolve, reject) => {
    ws.once('open',  resolve);
    ws.once('error', reject);
  });

  ws.on('message', raw => onMessage(raw, ws));
  ws.on('close',   (code, reason) =>
    console.log(`[${AGENT_ID}] Connection closed: ${code} – ${reason}`));

  // --- Tell Unity to follow this agent's camera ---
  await sleep(400);
  setView(ws);

  // --- Demonstrate all 6 controls ---

  console.log('\n--- Move forward ---');
  move(ws, 0, 1);
  await sleep(2000);

  console.log('\n--- Strafe right ---');
  move(ws, 1, 0);
  await sleep(1500);

  console.log('\n--- Sprint forward ---');
  move(ws, 0, 1);
  await sleep(2000);

  console.log('\n--- Stop ---');
  stop(ws);
  await sleep(500);

  console.log('\n--- Look right 90° ---');
  look(ws, 0, 90);
  await sleep(1000);

  console.log('\n--- Aim down sights ---');
  aim(ws, true);
  await sleep(1000);

  console.log('\n--- Shoot (0.5 s burst) ---');
  shoot(ws, true, 0.5);
  await sleep(1000);

  console.log('\n--- Reload ---');
  reloadWeapon(ws);
  await sleep(2500);

  console.log('\n--- Switch to slot 1 (sniper) ---');
  switchWeapon(ws, 1);
  await sleep(500);

  console.log('\n--- Switch back to slot 0 (rifle) ---');
  switchWeapon(ws, 0);
  await sleep(500);

  console.log('\n--- Exit ADS ---');
  aim(ws, false);

  console.log('\n--- Done. Stopping. ---');
  stop(ws);

  await sleep(500);
  ws.close();
}

main().catch(err => {
  console.error('Fatal error:', err.message);
  process.exit(1);
});

File: Test/python/single_agent.py
"""
single_agent.py  –  Use Case 1
================================
One script controls one character in the game.
The Unity display follows this agent's point-of-view.

Usage:
    python single_agent.py

Requirements:
    pip install websockets
"""

import asyncio
import json
import time
import websockets


UNITY_WS_URL = "ws://localhost:8080/agent"
AGENT_ID     = "agent_01"


# ---------------------------------------------------------------------------
# Low-level helpers
# ---------------------------------------------------------------------------

def make_command(command_type: str, data: dict, agent_id: str = AGENT_ID) -> str:
    return json.dumps({
        "commandType": command_type,
        "data":        data,
        "agentId":     agent_id,
        "timestamp":   time.time(),
    })


async def send(ws, command_type: str, data: dict = {}, agent_id: str = AGENT_ID):
    msg = make_command(command_type, data, agent_id)
    await ws.send(msg)
    print(f"[{agent_id}] → {command_type} {data}")


# ---------------------------------------------------------------------------
# High-level actions
# ---------------------------------------------------------------------------

async def move(ws, x: float, z: float):
    """Move: x = strafe, z = forward (normalized)."""
    await send(ws, "MOVE", {"x": x, "z": z})


async def stop(ws):
    await send(ws, "STOP", {})


async def look(ws, pitch: float, yaw: float):
    """Look to absolute pitch/yaw in degrees."""
    await send(ws, "LOOK", {"pitch": pitch, "yaw": yaw})


async def shoot(ws, active: bool = True, duration: float = 0):
    await send(ws, "SHOOT", {"active": active, "duration": duration})


async def reload_weapon(ws):
    await send(ws, "RELOAD", {})


async def switch_weapon(ws, slot: int):
    """slot is 0-based (0 = rifle, 1 = sniper, 2 = pistol …)."""
    await send(ws, "SWITCH_WEAPON", {"weaponIndex": slot})


async def aim(ws, active: bool):
    await send(ws, "AIM", {"active": active})


async def set_view(ws, target_agent_id: str = AGENT_ID):
    """Ask Unity to display this agent's camera."""
    await send(ws, "SET_VIEW", {"viewTargetAgentId": target_agent_id})


# ---------------------------------------------------------------------------
# Game-state listener
# ---------------------------------------------------------------------------

async def listen_state(ws):
    """Background task: print game state updates from Unity."""
    async for raw in ws:
        try:
            msg = json.loads(raw)
            if msg.get("type") == "welcome":
                print(f"[{AGENT_ID}] Connected – session {msg.get('sessionId')}")
                continue
            if "player" in msg:
                p = msg["player"]
                print(
                    f"[{AGENT_ID}] HP:{p.get('health',0):.0f}/{p.get('maxHealth',0):.0f} "
                    f"Ammo:{p.get('currentAmmo',0)}/{p.get('maxAmmo',0)} "
                    f"State:{p.get('movementState','?')} "
                    f"Enemies:{len(msg.get('enemies', []))}"
                )
        except Exception as e:
            print(f"[{AGENT_ID}] Parse error: {e}")


# ---------------------------------------------------------------------------
# Main scenario
# ---------------------------------------------------------------------------

async def main():
    print(f"[{AGENT_ID}] Connecting to {UNITY_WS_URL} ...")
    async with websockets.connect(UNITY_WS_URL) as ws:
        # Start listening to state updates in the background
        listener = asyncio.create_task(listen_state(ws))

        # --- Tell Unity to follow this agent's camera ---
        await asyncio.sleep(0.5)
        await set_view(ws)

        # --- Demonstrate all 6 controls ---

        print("\n--- Move forward ---")
        await move(ws, 0.0, 1.0)
        await asyncio.sleep(2.0)

        print("\n--- Strafe right ---")
        await move(ws, 1.0, 0.0)
        await asyncio.sleep(1.5)

        print("\n--- Sprint forward (full forward input) ---")
        await move(ws, 0.0, 1.0)
        await asyncio.sleep(2.0)

        print("\n--- Stop ---")
        await stop(ws)
        await asyncio.sleep(0.5)

        print("\n--- Look right 90° ---")
        await look(ws, 0.0, 90.0)
        await asyncio.sleep(1.0)

        print("\n--- Aim down sights ---")
        await aim(ws, True)
        await asyncio.sleep(1.0)

        print("\n--- Shoot (burst) ---")
        await shoot(ws, active=True, duration=0.5)
        await asyncio.sleep(1.0)

        print("\n--- Reload ---")
        await reload_weapon(ws)
        await asyncio.sleep(2.5)   # wait for reload animation

        print("\n--- Switch to slot 1 (sniper) ---")
        await switch_weapon(ws, 1)
        await asyncio.sleep(0.5)

        print("\n--- Switch back to slot 0 (rifle) ---")
        await switch_weapon(ws, 0)
        await asyncio.sleep(0.5)

        print("\n--- Exit ADS ---")
        await aim(ws, False)

        print("\n--- Done. Stopping. ---")
        await stop(ws)

        listener.cancel()


if __name__ == "__main__":
    asyncio.run(main())
